src/utils/fs.js

```diff
@@ -1,10 +1,29 @@
-const { readdir, unlink } = require('fs')
+const { lstat, readdir, readFile, unlink } = require('fs')
 const { format, join, parse, resolve } = require('path')
 const { promisify } = require('util')
 
+const pLstat = promisify(lstat)
 const pReaddir = promisify(readdir)
+const pReadFile = promisify(readFile)
 const pUnlink = promisify(unlink)
 
+// This caches multiple FS calls to the same path. It creates a cache key with
+// the name of the function and the path (e.g. "readdir:/some/directory").
+const cachedIOFunction = (func, cache, path, ...args) => {
```

Comment:
When are we expecting to get cache hits?

Aren't the paths passed to `findFunctionsInPaths` distinct?

Reply:
> Aren't the paths passed to findFunctionsInPaths distinct?

No. Each runtime will get the list of paths that have not been claimed by previous runtimes. So if you have a directory that doesn't contain any valid functions, it'll get passed to each runtime's `findFunctionsInPaths` and each runtime will try to read the directory and its contents to determine whether it contains a valid function, so that directory will go through a `stat` call (and a few other FS calls) for each runtime.

That's the duplication of work we're trying to avoid.

The JS runtime is not yet taking full advantage of this caching, because it's using `locate-path`, which we have no control over. I'll get to that in a future PR.

Reply:
> No. Each runtime will get the list of paths that have not been claimed by previous runtimes.

Cool, that makes sense. Thanks for clarifying